In [1]:
import tensorflow as tf
%matplotlib inline
import matplotlib as mpl
import matplotlib.pyplot as plt
import numpy as np
import os
import pandas as pd

mpl.rcParams['figure.figsize'] = (8, 6)
mpl.rcParams['axes.grid'] = False

# get train data (use for train & test )

# get test data

In [2]:
path = lambda number:r"C:\Users\14020\Desktop\NCBDC 2019\clearTestData1"+'\\'+str(number)+".csv"
testdflist=[]
DFSIZE=100
pd.read_csv(path(0))
for i in range(DFSIZE):
    df=pd.read_csv( path(i) ).iloc[:,1:]
    testdflist.append(df)

# get train data without charging

In [3]:
path = lambda number:r"C:\Users\14020\Desktop\NCBDC 2019\continueTimedataNC"+'\\'+str(number)+".csv"
NCdflist=[]
DFSIZE=419
for i in range(DFSIZE):
    df=pd.read_csv( path(i) ).iloc[:,1:]
    NCdflist.append(df)

## get the gradian of milediff/soc (once run)

## add some useless features (like tempdiff & SOCgap) (once run)

In [4]:
print(NCdflist[0].columns)

Index(['speed', 'total_voltage', 'total_current', 'soc', 'temp_max',
       'temp_min', 'motor_voltage', 'motor_current', 'mileage', 'milediff',
       'total_P', 'motor_P', 'speeddiff', 'milegap', 'tempMAXMINdiff',
       'SOCgap'],
      dtype='object')


# 多元线性回归
#### 筛选出 soc SOCgap temp_max temp_min tempMAXMINdiff 作为回归特征
#### milegap为回归量

#### 数据准备

In [5]:
df=NCdflist[0]
X=df.loc[:,['soc','SOCgap','temp_max','temp_min','tempMAXMINdiff']].to_numpy()
y=df.loc[:,['milegap']].to_numpy()

#### 模型训练

In [7]:
from sklearn.model_selection import train_test_split #这里是引用了交叉验证
from sklearn.linear_model import LinearRegression  #线性回归
import numpy as np
import matplotlib.pyplot as plt
X_train,X_test, y_train, y_test = train_test_split(X,y,test_size = 0.2,random_state=100)
linreg = LinearRegression()
model=linreg.fit(X_train, y_train)
print (model)
# 训练后模型截距
print("训练后模型截距",linreg.intercept_)
# 训练后模型权重（特征个数无变化）
print("训练后模型权重（特征个数无变化）",linreg.coef_)


LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)
训练后模型截距 [4.67511116e+13]
训练后模型权重（特征个数无变化） [[-4.67511116e+11 -4.67511116e+11 -2.42135705e+09  2.42135705e+09
   2.42135705e+09]]


#### 测试

In [8]:
#预测
y_pred = linreg.predict(X_test)
for i in range(len(y_pred)):
    print(i,y_test[i],y_pred[i])

0 [58.] [58.9609375]
1 [36.2] [35.609375]
2 [10.6] [11.1953125]
3 [70.3] [70.0234375]
4 [16.8] [17.0234375]
5 [30.5] [29.7890625]
6 [87.] [86.1796875]
7 [51.1] [51.2578125]
8 [97.6] [98.390625]
9 [85.] [84.4375]
10 [36.2] [35.609375]
11 [18.1] [18.765625]
12 [66.7] [68.1484375]
13 [88.1] [87.9296875]
14 [2.4] [0.7265625]
15 [70.3] [70.0234375]
16 [83.9] [84.4375]
17 [70.3] [70.4921875]
18 [96.] [94.8984375]
19 [34.4] [33.2734375]
20 [2.1] [0.59375]
21 [63.6] [64.6640625]
22 [70.3] [70.0234375]
23 [19.3] [20.375]
24 [79.7] [80.953125]
25 [65.2] [66.40625]
26 [19.3] [19.9140625]
27 [36.2] [35.609375]
28 [70.3] [69.890625]
29 [11.7] [11.65625]
30 [96.3] [95.03125]
31 [36.2] [35.1484375]
32 [36.2] [35.609375]
33 [80.5] [80.953125]
34 [52.2] [53.]
35 [19.3] [19.78125]
36 [64.5] [64.6640625]
37 [19.3] [20.375]
38 [70.3] [70.4921875]
39 [46.9] [47.3046875]
40 [7.4] [7.7109375]
41 [2.4] [0.59375]
42 [100.] [99.5390625]
43 [98.5] [98.390625]
44 [36.2] [35.7421875]
45 [36.7] [37.4921875]
46 [70.

### 用上面的给出的回归模型给所有的数据一个新特征 （linearMilegap)

In [10]:
linearModel=model
LinearModel_X = lambda df:df.loc[:,['soc','SOCgap','temp_max','temp_min','tempMAXMINdiff']].to_numpy()
LinearModel_y = lambda df:df.loc[:,['milegap']].to_numpy()
df=NCdflist[3]
X_test=LinearModel_X(df)
y_test=LinearModel_y(df)
y_pred=linearModel.predict(LinearModel_X(df))
for i in range(len(y_pred)):
    print(i,y_test[i],y_pred[i])

0 [0.] [2.2109375]
1 [0.] [2.2109375]
2 [0.] [2.2109375]
3 [0.] [2.8125]
4 [0.1] [2.8125]
5 [0.1] [2.8125]
6 [0.1] [2.8125]
7 [0.1] [2.8125]
8 [0.1] [2.8125]
9 [0.1] [2.8125]
10 [0.1] [2.8125]
11 [0.1] [2.8125]
12 [0.1] [2.8125]
13 [0.1] [2.8125]
14 [0.1] [2.8125]
15 [0.1] [2.8125]
16 [0.1] [2.8125]
17 [0.2] [4.5546875]
18 [0.3] [4.5546875]
19 [0.4] [4.5546875]
20 [0.5] [4.5546875]
21 [0.6] [4.4296875]
22 [0.7] [4.4296875]
23 [0.8] [4.4296875]
24 [0.9] [4.4296875]
25 [0.9] [4.4296875]
26 [0.9] [4.4296875]
27 [0.9] [4.4296875]
28 [1.] [6.171875]
29 [1.2] [6.171875]
30 [1.3] [6.171875]
31 [1.4] [6.171875]
32 [1.5] [6.171875]
33 [1.7] [6.171875]
34 [1.8] [6.171875]
35 [1.9] [6.171875]
36 [2.] [6.171875]
37 [2.1] [6.171875]
38 [2.2] [6.171875]
39 [2.3] [6.171875]
40 [2.4] [6.171875]
41 [2.4] [6.171875]
42 [2.4] [6.171875]
43 [2.4] [6.171875]
44 [2.4] [6.171875]
45 [2.4] [7.921875]
46 [2.5] [7.453125]
47 [2.6] [7.453125]
48 [2.6] [7.453125]
49 [2.6] [7.453125]
50 [2.6] [7.453125]
51 [2.7] [

679 [33.7] [36.625]
680 [33.8] [36.625]
681 [33.9] [36.625]
682 [34.] [38.375]
683 [34.1] [38.375]
684 [34.3] [38.375]
685 [34.4] [38.375]
686 [34.5] [38.375]
687 [34.6] [38.375]
688 [34.8] [38.375]
689 [34.9] [38.375]
690 [35.05] [38.375]
691 [35.1] [38.375]
692 [35.2] [38.375]
693 [35.3] [38.375]
694 [35.4] [38.375]
695 [35.5] [38.375]
696 [35.6] [38.375]
697 [35.7] [38.375]
698 [35.8] [38.375]
699 [35.9] [38.375]
700 [36.] [38.375]
701 [36.1] [38.375]
702 [36.2] [38.375]
703 [36.3] [38.375]
704 [36.4] [38.375]
705 [36.6] [38.375]
706 [36.7] [38.375]
707 [36.8] [38.375]
708 [36.9] [38.375]
709 [37.1] [40.109375]
710 [37.2] [40.109375]
711 [37.3] [40.109375]
712 [37.4] [40.109375]
713 [37.5] [40.109375]
714 [37.6] [40.109375]
715 [37.7] [40.109375]
716 [37.8] [40.109375]
717 [37.8] [41.8515625]
718 [37.8] [41.8515625]
719 [37.8] [41.8515625]
720 [37.8] [41.8515625]
721 [37.8] [41.8515625]
722 [37.8] [41.8515625]
723 [37.8] [41.8515625]
724 [37.8] [41.8515625]
725 [37.8] [41.8515625]
7

1305 [65.2] [68.3984375]
1306 [65.2] [68.3984375]
1307 [65.2] [68.3984375]
1308 [65.3] [68.3984375]
1309 [65.4] [68.3984375]
1310 [65.5] [68.3984375]
1311 [65.5] [68.3984375]
1312 [65.5] [68.3984375]
1313 [65.5] [68.3984375]
1314 [65.5] [68.3984375]
1315 [65.5] [68.3984375]
1316 [65.5] [68.3984375]
1317 [65.55] [68.3984375]
1318 [65.5] [68.3984375]
1319 [65.6] [68.3984375]
1320 [65.6] [68.3984375]
1321 [65.7] [68.3984375]
1322 [65.7] [68.3984375]
1323 [65.7] [68.3984375]
1324 [65.85] [69.265625]
1325 [65.9] [70.1328125]
1326 [66.] [70.1328125]
1327 [66.1] [70.1328125]
1328 [66.2] [70.1328125]
1329 [66.2] [70.1328125]
1330 [66.3] [70.1328125]
1331 [66.4] [70.1328125]
1332 [66.5] [70.1328125]
1333 [66.5] [70.1328125]
1334 [66.5] [70.1328125]
1335 [66.5] [70.1328125]
1336 [66.5] [70.1328125]
1337 [66.5] [70.1328125]
1338 [66.6] [70.1328125]
1339 [66.6] [70.1328125]
1340 [66.7] [70.1328125]
1341 [66.7] [70.1328125]
1342 [66.8] [70.1328125]
1343 [66.9] [70.1328125]
1344 [67.] [70.1328125]
1

1808 [89.3] [90.6875]
1809 [89.3] [90.6875]
1810 [89.3] [90.6875]
1811 [89.3] [90.6875]
1812 [89.3] [90.6875]
1813 [89.3] [90.6875]
1814 [89.3] [90.6875]
1815 [89.3] [90.6875]
1816 [89.3] [90.6875]
1817 [89.3] [90.6875]
1818 [89.3] [90.6875]
1819 [89.3] [90.6875]
1820 [89.3] [90.6875]
1821 [89.3] [90.6875]
1822 [89.3] [90.6875]
1823 [89.3] [90.6875]
1824 [89.3] [90.6875]
1825 [89.3] [90.6875]
1826 [89.3] [90.6875]
1827 [89.3] [90.6875]
1828 [89.3] [90.6875]
1829 [89.3] [90.6875]
1830 [89.3] [90.6875]
1831 [89.3] [90.6875]
1832 [89.3] [92.421875]
1833 [89.3] [92.421875]
1834 [89.3] [92.421875]
1835 [89.4] [92.421875]
1836 [89.4] [92.421875]
1837 [89.5] [92.421875]
1838 [89.5] [92.421875]
1839 [89.5] [92.4296875]
1840 [89.5] [92.421875]
1841 [89.5] [92.421875]
1842 [89.6] [92.421875]
1843 [89.6] [92.421875]
1844 [89.6] [92.421875]
1845 [89.7] [92.421875]
1846 [89.8] [92.421875]
1847 [89.9] [92.421875]
1848 [90.] [92.421875]
1849 [90.1] [92.421875]
1850 [90.2] [92.421875]
1851 [90.3] [92.

## 调用此函数获得线性模型的输出 array（可作为特征）

In [11]:
def LinearResult(model,df):
    X_test=df.loc[:,['soc','SOCgap','temp_max','temp_min','tempMAXMINdiff']].to_numpy()
    y_test=df.loc[:,['milegap']].to_numpy()
    y_pred=model.predict(X_test)
    return y_pred,y_test

# GBDT回归残差

#### 首先得到线性回归的预测值，然后计算残差 mileResidual

In [12]:
mileResidualList=[]
for i in NCdflist:
    y_pred,y_test=LinearResult(model,i)
    mileResidualList.append(y_pred-y_test)

In [20]:
for i in range(len(mileResidualList)):
    print(mileResidualList[i][0])

[-2.8984375]
[2.2109375]
[-1.234375]
[2.2109375]
[-1.8359375]
[3.08557337e+13]
[2.2109375]
[-1.375]
[2.8125]
[-4.6875]
[3.875]
[-7.8671875]
[-6.546875]
[2.05704891e+13]
[-1.109375]
[2.24405336e+13]
[-4.6875]
[1.68304002e+13]
[-3.1640625]
[2.10380002e+13]
[4.67511116e+11]
[-3.953125]
[-0.640625]
[1.015625]
[2.05704891e+13]
[-1.8359375]
[2.8125]
[0.5546875]
[2.8050667e+13]
[2.33755558e+12]
[2.0102978e+13]
[-2.3046875]
[4.2734375]
[2.4765625]
[-3.953125]
[2.9453125]
[-0.0390625]
[-9.6640625]
[-12.8515625]
[9.35022232e+11]
[-13.4453125]
[-11.921875]
[-3.1640625]
[-0.7109375]
[-4.953125]
[-1.765625]
[1.15625]
[-1.8359375]
[2.2109375]
[1.16877779e+13]
[-1.109375]
[2.6796875]
[-1.109375]
[-3.2578125]
[-2.8984375]
[2.3515625]
[0.5546875]
[2.33755558e+12]
[-0.5078125]
[2.33755558e+13]
[2.3515625]
[-5.1484375]
[1.484375]
[-2.4296875]
[2.15055113e+13]
[-2.8984375]
[-1.8359375]
[1.015625]
[2.29080447e+13]
[-2.8984375]
[2.47780892e+13]
[1.484375]
[-2.3046875]
[2.61806225e+13]
[4.140625]
[3.03882226

#### 观察结果是否有大偏差

In [22]:
for index,i in enumerate(mileResidualList):
    if(abs(i[-1])>1000):
        print(index,i[0],len(NCdflist[index]))

5 [3.08557337e+13] 25
13 [2.05704891e+13] 528
15 [2.24405336e+13] 562
17 [1.68304002e+13] 906
19 [2.10380002e+13] 670
20 [4.67511116e+11] 2054
24 [2.05704891e+13] 684
28 [2.8050667e+13] 346
29 [2.33755558e+12] 1
30 [2.0102978e+13] 975
39 [9.35022232e+11] 1786
49 [1.16877779e+13] 1683
57 [2.33755558e+12] 2220
59 [2.33755558e+13] 454
64 [2.15055113e+13] 699
68 [2.29080447e+13] 616
70 [2.47780892e+13] 304
73 [2.61806225e+13] 251
75 [3.03882226e+13] 395
77 [3.31932892e+13] 297
78 [1.87004446e+12] 1641
83 [2.4310578e+13] 367
89 [4.67511116e+11] 2694
93 [1.2155289e+13] 1277
155 [2.94532003e+13] 64
157 [1.40253335e+12] 1859
160 [2.89856892e+13] 188
162 [2.89856892e+13] 42
166 [1.40253335e+13] 1413
171 [3.13232448e+13] 43
176 [1.49603557e+13] 791
181 [3.50633337e+13] 66
184 [3.45958226e+13] 24
186 [1.77654224e+13] 524
199 [1.68304002e+13] 597
214 [1.07527557e+13] 1290
219 [6.07764451e+12] 1214
258 [3.64658671e+13] 19
265 [1.30903113e+13] 598
270 [3.08557337e+13] 59
272 [1.12202668e+13] 1129
27

In [24]:
#5 13  15 17 19 20 24 28 29 30  39 这些数据集 偏差较大 
y_pred,y_test=LinearResult(model,NCdflist[157])

In [26]:
y_pred

array([[1.40253335e+12],
       [1.40253335e+12],
       [1.40253335e+12],
       ...,
       [1.40253335e+12],
       [1.40253335e+12],
       [1.40253335e+12]])

#### 将mileResidual写入列
### 然后将数据分成训练集和测试集 每四个数据前三个加入训练集 最后一个作为测试集

In [399]:
getMileResidual = lambda number:mileResidualList[number]
trainlist,testlist=[],[]
for index,i in enumerate(NCdflist):
    i['mileResidual']=getMileResidual(index)
    if index%4==0 or index%4==1 or index%4==2:
        trainlist.append(i)
    else:
        testlist.append(i)

### 然后将训练集的数据准备好
#### （暂时不考虑speed和其差值）
##### 输入数据特征为 'total_voltage', 'total_current','motor_voltage', 'motor_current', 'total_P', 'motor_P'|||||||||||| 'speeddiff'
##### 输出特征为mileResidual,||||||||||speed

In [417]:
GBDT_X_DF = lambda df:df.loc[:,['total_voltage', 'total_current','motor_voltage', 'motor_current', 'total_P', 'motor_P']]
GBDT_y_DF = lambda df:df.loc[:,['mileResidual']]
df=trainlist[0]
X_train=GBDT_X_DF(df)
y_train=GBDT_y_DF(df)

###### GDBT的数据归一化 只能用训练集的方差和均值

In [420]:
from sklearn.preprocessing import StandardScaler
XScaler = StandardScaler().fit(X_train)  
YScaler = StandardScaler().fit(y_train)  
X=XScaler.transform(X_train)
y=YScaler.transform(y_train)

In [429]:
x_train, x_test, y_train, y_test = train_test_split(X,y,test_size=0.1, random_state=0)
X.shape,y.shape

((1911, 6), (1911, 1))

In [430]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import mean_squared_error
from sklearn import ensemble
from sklearn.model_selection import train_test_split
#myX, myy, _X , _y = train_test_split(X,y,test_size=0.1, random_state=0)
params = {'n_estimators': 250, 'max_depth': 3, 'min_samples_split': 2,
              'learning_rate': 0.01, 'loss': 'ls'}
clf = ensemble.GradientBoostingRegressor(**params)
clf.fit(x_train, y_train)

D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\ensemble\gradient_boosting.py:1450: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


GradientBoostingRegressor(alpha=0.9, criterion='friedman_mse', init=None,
                          learning_rate=0.01, loss='ls', max_depth=3,
                          max_features=None, max_leaf_nodes=None,
                          min_impurity_decrease=0.0, min_impurity_split=None,
                          min_samples_leaf=1, min_samples_split=2,
                          min_weight_fraction_leaf=0.0, n_estimators=250,
                          n_iter_no_change=None, presort='auto',
                          random_state=None, subsample=1.0, tol=0.0001,
                          validation_fraction=0.1, verbose=0, warm_start=False)

### 准备测试数据

In [431]:
df=testlist[0]
X_test=GBDT_X_DF(df)
y_test=GBDT_y_DF(df).to_numpy()
X=XScaler.transform(X_test)
y=YScaler.transform(y_test)
y_predict=clf.predict(X_test)
mse = mean_squared_error(y_test, y_predict)
print("MSE: %.4f" % mse)

MSE: 27.4629


### 第一次：用第一个训练集数据训练 用第一个测试集数据测试 MSE: 24.7331（忘记加train_test_split）
### 第二次：用第一个训练集数据训练 用第一个测试集数据测试 MSE: 27.4629（加train_test_split）

In [356]:
for i in range(len(y_test)):
    print(i,y_test[i],y_predict[i])

0 [2.2109375] -1.864747219718846
1 [2.2109375] -2.5719779595986063
2 [2.2109375] -1.864747219718846
3 [2.8125] -1.864747219718846
4 [2.7125] -2.5719779595986063
5 [2.7125] -1.864747219718846
6 [2.7125] -1.864747219718846
7 [2.7125] -1.864747219718846
8 [2.7125] -1.864747219718846
9 [2.7125] -1.864747219718846
10 [2.7125] -1.864747219718846
11 [2.7125] -1.864747219718846
12 [2.7125] -1.864747219718846
13 [2.7125] -1.864747219718846
14 [2.7125] -1.864747219718846
15 [2.7125] -2.5719779595986063
16 [2.7125] -1.864747219718846
17 [4.3546875] -2.5719779595986063
18 [4.2546875] -2.5719779595986063
19 [4.1546875] -2.5719779595986063
20 [4.0546875] -2.5719779595986063
21 [3.8296875] -2.5719779595986063
22 [3.7296875] -2.5719779595986063
23 [3.6296875] -1.864747219718846
24 [3.5296875] -1.864747219718846
25 [3.5296875] -1.864747219718846
26 [3.5296875] -2.5719779595986063
27 [3.5296875] -2.5719779595986063
28 [5.171875] -2.5719779595986063
29 [4.971875] -2.5719779595986063
30 [4.871875] -2.5719

314 [2.084375] -1.5679045735408579
315 [2.084375] -1.5679045735408579
316 [2.084375] -1.5679045735408579
317 [2.084375] -1.5679045735408579
318 [2.084375] -1.5679045735408579
319 [2.084375] -1.5679045735408579
320 [2.084375] -1.5679045735408579
321 [2.084375] -1.5679045735408579
322 [2.084375] -1.5679045735408579
323 [2.084375] -1.5679045735408579
324 [2.084375] -1.5679045735408579
325 [2.084375] -1.5679045735408579
326 [2.084375] -1.5679045735408579
327 [2.084375] -1.5679045735408579
328 [2.084375] -1.5679045735408579
329 [2.084375] -1.5679045735408579
330 [2.084375] -1.5679045735408579
331 [2.084375] -1.5679045735408579
332 [2.084375] -1.5679045735408579
333 [2.084375] -1.5679045735408579
334 [2.084375] -1.5679045735408579
335 [2.084375] -1.5679045735408579
336 [2.084375] -1.5679045735408579
337 [2.084375] -1.5679045735408579
338 [2.084375] -1.5679045735408579
339 [2.084375] -1.5679045735408579
340 [2.084375] -1.5679045735408579
341 [2.084375] -1.5679045735408579
342 [2.084375] -1.56

979 [4.0578125] -0.728259793756717
980 [4.0578125] -1.6787618557425592
981 [4.0578125] -1.6787618557425592
982 [4.0578125] -1.6787618557425592
983 [4.0578125] -1.6787618557425592
984 [4.0578125] -1.6787618557425592
985 [4.0578125] -2.5719779595986063
986 [3.9578125] -2.5719779595986063
987 [3.8578125] -2.5719779595986063
988 [3.7578125] -2.560232210707855
989 [3.6578125] -0.728259793756717
990 [3.5578125] -2.5719779595986063
991 [3.5578125] -1.5679045735408579
992 [3.5578125] -1.6787618557425592
993 [3.5578125] -1.6787618557425592
994 [3.5578125] -1.6787618557425592
995 [3.6984375] -1.6787618557425592
996 [3.6984375] -1.6787618557425592
997 [3.6984375] -1.6787618557425592
998 [3.5984375] -2.5719779595986063
999 [3.4984375] -2.5719779595986063
1000 [3.4984375] -0.728259793756717
1001 [3.3984375] -2.5719779595986063
1002 [3.2984375] -2.5719779595986063
1003 [3.1984375] -0.728259793756717
1004 [3.0984375] -0.728259793756717
1005 [2.9984375] -2.5719779595986063
1006 [2.9984375] -2.57197795

1484 [3.4671875] -1.5679045735408579
1485 [3.4671875] -1.5679045735408579
1486 [3.4671875] -1.5679045735408579
1487 [3.4671875] -1.5679045735408579
1488 [3.4671875] -1.5679045735408579
1489 [3.4671875] -1.5679045735408579
1490 [3.4671875] -1.5679045735408579
1491 [3.4671875] -1.5679045735408579
1492 [3.4671875] -1.5679045735408579
1493 [3.4671875] -1.5679045735408579
1494 [3.4671875] -1.5679045735408579
1495 [3.4671875] -1.5679045735408579
1496 [3.4671875] -1.5679045735408579
1497 [3.4671875] -1.5679045735408579
1498 [3.4671875] -1.5679045735408579
1499 [3.4671875] -2.5719779595986063
1500 [3.4671875] -2.5719779595986063
1501 [3.4671875] -2.5719779595986063
1502 [3.3671875] -2.407590796351333
1503 [3.2671875] -1.5888625838985713
1504 [3.2671875] -0.728259793756717
1505 [3.2671875] -1.6787618557425592
1506 [5.0171875] -2.5719779595986063
1507 [4.9171875] -2.5719779595986063
1508 [4.7171875] -1.5888625838985713
1509 [4.6171875] -2.5719779595986063
1510 [4.5171875] -2.5719779595986063
151

2147 [2.5515625] -1.5679045735408579
2148 [2.5515625] -1.5679045735408579
2149 [2.5515625] -1.5679045735408579
2150 [2.5515625] -1.5679045735408579
2151 [2.5515625] -1.5679045735408579
2152 [2.5515625] -1.5679045735408579
2153 [2.5515625] -1.5679045735408579
2154 [2.5515625] -1.5679045735408579
2155 [2.5515625] -1.5679045735408579
2156 [2.5515625] -1.5679045735408579
2157 [2.5515625] -1.5679045735408579
2158 [2.5515625] -1.5679045735408579
2159 [2.5515625] -1.5679045735408579
2160 [2.5515625] -1.5679045735408579
2161 [2.5515625] -1.5679045735408579
2162 [2.5515625] -1.5679045735408579
2163 [2.5515625] -1.5679045735408579
2164 [2.5515625] -1.5679045735408579
2165 [2.5515625] -1.5679045735408579
2166 [2.5515625] -1.5679045735408579
2167 [2.5515625] -1.5679045735408579
2168 [2.5515625] -1.5679045735408579
2169 [2.5515625] -1.5679045735408579
2170 [2.5515625] -1.5679045735408579
2171 [2.5515625] -1.5679045735408579
2172 [2.5515625] -1.5679045735408579
2173 [2.5515625] -1.5679045735408579
2

## 上面训练集只用了一段 貌似效果不佳 下面是
# 用上所有训练数据的GDBT

### 首先获得所有数据的归一化
##### 获得所有数据

In [367]:
alldata=pd.DataFrame()
for df in trainlist:
    alldata=pd.concat([alldata,df])

#### 创建归一化器

In [369]:
from sklearn.preprocessing import StandardScaler
X_all=GBDT_X_DF(alldata)
y_all=GBDT_y_DF(alldata)
XScaler = StandardScaler().fit(X_all)
YScaler = StandardScaler().fit(y_all)

In [370]:
#其实这里如果用增量学习器也不错，etc.SGDRegressor PassiveAggressiveRegressor
#more detail:https://blog.csdn.net/whiterbear/article/details/53120004
params = {'n_estimators': 250, 'max_depth': 3, 'min_samples_split': 2,
              'learning_rate': 0.01, 'loss': 'ls'}
clf = ensemble.GradientBoostingRegressor(**params)
X_train=X_all.to_numpy()
y_train=y_all.to_numpy()

X=XScaler.transform(X_train)
y=YScaler.transform(y_train)

clf.fit(X, y)

D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\ensemble\gradient_boosting.py:1450: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


GradientBoostingRegressor(alpha=0.9, criterion='friedman_mse', init=None,
                          learning_rate=0.01, loss='ls', max_depth=3,
                          max_features=None, max_leaf_nodes=None,
                          min_impurity_decrease=0.0, min_impurity_split=None,
                          min_samples_leaf=1, min_samples_split=2,
                          min_weight_fraction_leaf=0.0, n_estimators=250,
                          n_iter_no_change=None, presort='auto',
                          random_state=None, subsample=1.0, tol=0.0001,
                          validation_fraction=0.1, verbose=0, warm_start=False)

## 测试数据

In [371]:
df=testlist[0]
X_test=GBDT_X_DF(df)
y_test=GBDT_y_DF(df).to_numpy()
X=XScaler.transform(X_test)
y=YScaler.transform(y_test)
y_predict=clf.predict(X_test)
mse = mean_squared_error(y_test, y_predict)
print("MSE: %.4f" % mse)

MSE: 11.0215


### 之前是MSE: 24.7331  现在是MSE: 11.0215，数据加大还是有一些效果的

In [373]:
for i in range(len(X_test)):
    print(i,y_test[i],y_predict[i])

0 [2.2109375] -0.20685211849140187
1 [2.2109375] -0.21126451414733452
2 [2.2109375] -0.20685211849140187
3 [2.8125] -0.20685211849140187
4 [2.7125] -0.21126451414733452
5 [2.7125] -0.20685211849140187
6 [2.7125] -0.20685211849140187
7 [2.7125] -0.20685211849140187
8 [2.7125] -0.20685211849140187
9 [2.7125] -0.20685211849140187
10 [2.7125] -0.20685211849140187
11 [2.7125] -0.20685211849140187
12 [2.7125] -0.20685211849140187
13 [2.7125] -0.20685211849140187
14 [2.7125] -0.20685211849140187
15 [2.7125] -0.21126451414733452
16 [2.7125] -0.20685211849140187
17 [4.3546875] -0.21126451414733452
18 [4.2546875] -0.21126451414733452
19 [4.1546875] -0.21126451414733452
20 [4.0546875] -0.21126451414733452
21 [3.8296875] -0.21126451414733452
22 [3.7296875] -0.21126451414733452
23 [3.6296875] -0.20685211849140187
24 [3.5296875] -0.20685211849140187
25 [3.5296875] -0.20685211849140187
26 [3.5296875] -0.21126451414733452
27 [3.5296875] -0.21126451414733452
28 [5.171875] -0.21126451414733452
29 [4.971

308 [2.084375] -0.18846044365771364
309 [2.084375] -0.18846044365771364
310 [2.084375] -0.18846044365771364
311 [2.084375] -0.18846044365771364
312 [2.084375] -0.18846044365771364
313 [2.084375] -0.18846044365771364
314 [2.084375] -0.18846044365771364
315 [2.084375] -0.18846044365771364
316 [2.084375] -0.18846044365771364
317 [2.084375] -0.18846044365771364
318 [2.084375] -0.18846044365771364
319 [2.084375] -0.18846044365771364
320 [2.084375] -0.18846044365771364
321 [2.084375] -0.18846044365771364
322 [2.084375] -0.18846044365771364
323 [2.084375] -0.18846044365771364
324 [2.084375] -0.18846044365771364
325 [2.084375] -0.18846044365771364
326 [2.084375] -0.18846044365771364
327 [2.084375] -0.18846044365771364
328 [2.084375] -0.18846044365771364
329 [2.084375] -0.18846044365771364
330 [2.084375] -0.18846044365771364
331 [2.084375] -0.18846044365771364
332 [2.084375] -0.18846044365771364
333 [2.084375] -0.18846044365771364
334 [2.084375] -0.18846044365771364
335 [2.084375] -0.1884604436

856 [4.5203125] -0.18846044365771364
857 [4.5203125] -0.18846044365771364
858 [4.5203125] -0.18846044365771364
859 [4.5203125] -0.18846044365771364
860 [4.5203125] -0.18846044365771364
861 [4.5203125] -0.18846044365771364
862 [4.5203125] -0.18846044365771364
863 [4.5203125] -0.18846044365771364
864 [4.5203125] -0.18846044365771364
865 [4.5203125] -0.18846044365771364
866 [4.5203125] -0.18846044365771364
867 [4.5203125] -0.18846044365771364
868 [4.5203125] -0.18846044365771364
869 [4.5203125] -0.18846044365771364
870 [4.5203125] -0.18846044365771364
871 [4.5203125] -0.18846044365771364
872 [4.5203125] -0.18846044365771364
873 [4.5203125] -0.18846044365771364
874 [4.5203125] -0.18846044365771364
875 [4.5203125] -0.18846044365771364
876 [4.5203125] -0.18846044365771364
877 [4.5203125] -0.18846044365771364
878 [4.5203125] -0.18846044365771364
879 [4.5203125] -0.18846044365771364
880 [4.5203125] -0.18846044365771364
881 [4.5203125] -0.18846044365771364
882 [4.4703125] -0.21126451414733452
8

1528 [4.159375] -0.21126451414733452
1529 [3.959375] -0.21126451414733452
1530 [3.859375] -0.11167946746900644
1531 [3.759375] -0.21126451414733452
1532 [3.659375] -0.20417171319333902
1533 [3.509375] -0.20685211849140187
1534 [3.459375] -0.21126451414733452
1535 [3.359375] -0.21126451414733452
1536 [3.4] -0.11167946746900644
1537 [3.4] -0.20685211849140187
1538 [3.3] -0.21126451414733452
1539 [3.3] -0.21126451414733452
1540 [3.2] -0.21126451414733452
1541 [4.734375] -0.21126451414733452
1542 [4.634375] -0.1857800383596508
1543 [4.634375] -0.21126451414733452
1544 [4.534375] -0.21126451414733452
1545 [4.334375] -0.21126451414733452
1546 [4.234375] -0.21126451414733452
1547 [4.134375] -0.11167946746900644
1548 [4.034375] -0.11167946746900644
1549 [3.9015625] -0.21126451414733452
1550 [3.934375] -0.21126451414733452
1551 [3.7015625] -0.21126451414733452
1552 [3.6015625] -0.11167946746900644
1553 [3.5015625] -0.21126451414733452
1554 [3.3015625] -0.21126451414733452
1555 [3.2015625] -0.21

2177 [2.5515625] -0.18846044365771364
2178 [2.5515625] -0.18846044365771364
2179 [2.5515625] -0.18846044365771364
2180 [2.5515625] -0.18846044365771364
2181 [2.5515625] -0.18846044365771364
2182 [2.5515625] -0.18846044365771364
2183 [2.5515625] -0.18846044365771364
2184 [2.5515625] -0.18846044365771364
2185 [2.5515625] -0.18846044365771364
2186 [2.5515625] -0.18846044365771364
2187 [2.5515625] -0.18846044365771364
2188 [2.5515625] -0.18846044365771364
2189 [2.5515625] -0.18846044365771364
2190 [2.5515625] -0.18846044365771364
2191 [2.5515625] -0.18846044365771364
2192 [2.5515625] -0.18846044365771364
2193 [2.5515625] -0.18846044365771364
2194 [2.5515625] -0.18846044365771364
2195 [2.5515625] -0.18846044365771364
2196 [2.5515625] -0.18846044365771364
2197 [2.5515625] -0.18846044365771364
2198 [2.5515625] -0.18846044365771364
2199 [2.5515625] -0.18846044365771364
2200 [2.5515625] -0.18846044365771364
2201 [2.5515625] -0.18846044365771364
2202 [2.5515625] -0.18846044365771364
2203 [2.5515

### 看这个数据感觉是因为LR没有做清楚，所以导致误差数据不好，影响到了残差的回归
### 还有一点就是我的训练集貌似没有随机抽样 
# 那么重新训练一个SGDRegressor
#### 创建迭代器：

In [435]:
LinearModel_X = lambda df:df.loc[:,['soc','SOCgap','temp_max','temp_min','tempMAXMINdiff']].to_numpy()
LinearModel_y = lambda df:df.loc[:,['milegap']].to_numpy()
def iter_minibatches(dflist):
    '''
    迭代器
    给定文件流（比如一个大文件），每次输出minibatch_size行，默认选择1k行
    将输出转化成numpy输出，返回X, y
    '''
    for df in dflist:
        X=LinearModel_X(df)
        y=LinearModel_y(df)
        yield X, y
# 生成测试文件
minibatch_test_iterators = iter_minibatches(NCdflist)
X_test, y_test = minibatch_test_iterators.__next__()  # 得到一份测试文件

### SGDRegreessor的归一化

In [444]:
from sklearn.preprocessing import StandardScaler
X_all=LinearModel_X(alldata)
y_all=LinearModel_y(alldata)
SGD_XScaler = StandardScaler().fit(X_all)
SGD_YScaler = StandardScaler().fit(y_all)

## 增量学习

In [448]:
from sklearn.linear_model import SGDRegressor
sgd = SGDRegressor()  # SGDClassifier的参数设置可以参考sklearn官网
minibatch_train_iterators = iter_minibatches(NCdflist)
for i, (X_train, y_train) in enumerate(minibatch_train_iterators):
    X_train=SGD_XScaler.transform(X_train)
    y_train=SGD_YScaler.transform(y_train)
    sgd.partial_fit(X_train, y_train)
    print("{} time".format(i))  # 当前次数
    print("{} score".format(sgd.score(X_test, y_test)))  # 在测试集上看效果

D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_

0 time
-5.214500485377001 score
1 time
-5.606757959537277 score
2 time
-5.3802641296273555 score
3 time
-5.590590083373221 score
4 time
-5.471866113250076 score
5 time
-4.420707944572264 score
6 time
-4.531019877511932 score
7 time
-4.635088857864407 score
8 time
-4.576910082532205 score
9 time
-4.461307720480328 score
10 time
-4.517832107566521 score
11 time
-4.274967108267573 score
12 time
-4.330375141089654 score
13 time
-2.103731492016326 score
14 time
-2.0871812396212834 score
15 time
-1.054263023201015 score
16 time
-1.0608206149671293 score
17 time
-0.6360077545432878 score
18 time
-0.6268129234831168 score
19 time
-0.5829080424435824 score
20 time
-0.6998316008907566 score
21 time
-0.9671068739813885 score
22 time
-0.8742821381959843 score
23 time
-0.8533310992122818 score
24 time
-0.36929358745162344 score
25 time
-0.42193056781101923 score
26 time
-0.36824243289290526 score
27 time
-0.347970470236594 score
28 time
-0.32278653646516187 score
29 time
-0.3217892798135249 score
3

D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_

41 time
-0.41469701901029676 score
42 time
-0.652015042050438 score
43 time
-0.9451254024718908 score
44 time
-0.8543572648743589 score
45 time
-0.4019498099119109 score
46 time
-0.4013263795852424 score
47 time
-0.3579873809560097 score
48 time
-0.36444261947728274 score
49 time
-0.32162190560253334 score
50 time
-0.294609311718697 score
51 time
-0.3289804926877029 score
52 time
-0.31277205347762616 score
53 time
-0.3438487476435914 score
54 time
-0.3163702706915166 score
55 time
-0.3088280847912015 score
56 time
-0.2950345171254245 score
57 time
-0.3116079098155973 score
58 time
-0.3147912688011123 score
59 time
-0.3483494257730699 score
60 time
-0.46378700018013785 score
61 time
-0.41023521769301396 score
62 time
-0.5005476380155971 score
63 time
-0.4441859790871838 score
64 time
-0.5387417423929888 score
65 time
-0.48825805766700264 score
66 time
-0.441103802533539 score
67 time
-0.48762405244240137 score
68 time
-0.38922282522971585 score
69 time
-0.3861828383170838 score
70 time


D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_

D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_


127 time
-0.13414959039742702 score
128 time
-0.23885956674627673 score
129 time
-0.17678849810813957 score
130 time
-0.1835416704239561 score
131 time
-0.1122652455847115 score
132 time
-0.18298129918880116 score
133 time
-0.22502255174733163 score
134 time
-0.3078487531304823 score
135 time
-0.21715984259316445 score
136 time
-0.22753228665486458 score
137 time
-0.2082583225510792 score
138 time
-0.16354435058596328 score
139 time
-0.12447212110699057 score
140 time
-0.3094173992589917 score
141 time
-0.20223424823701452 score
142 time
-0.212642857336649 score
143 time
-0.19671885829677205 score
144 time
-0.14857821642616598 score
145 time
-0.11916005676639818 score
146 time
-0.07607861767951185 score
147 time
-0.030762614184444814 score
148 time
-0.14263191192129443 score
149 time
-0.1387417211680162 score
150 time
-0.023605481821866284 score
151 time
-0.07477441439294275 score
152 time
-0.09130690412907728 score
153 time
-0.08749022737574252 score
154 time
-0.09131004635591866 sco

D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_

168 time
-0.7376021173482129 score
169 time
-0.7764497164259443 score
170 time
-0.8861643746091461 score
171 time
-0.7848253398633253 score
172 time
-0.8045636709251657 score
173 time
-0.6731332068593785 score
174 time
-0.7406906619331843 score
175 time
-0.7036668546680762 score
176 time
-0.5990024707847534 score
177 time
-0.5439277049163187 score
178 time
-0.5244578870982344 score
179 time
-0.486883222437984 score
180 time
-0.4853062369516974 score
181 time
-0.4159483020197939 score
182 time
-0.452657792348365 score
183 time
-0.48500984064210906 score
184 time
-0.4655611768845047 score
185 time
-0.46080163776915684 score
186 time
-0.5218558536452826 score
187 time
-0.5099056732037546 score
188 time
-0.49834520365252355 score
189 time
-0.5889406221735214 score
190 time
-0.5854764319261117 score
191 time
-0.5581343394202962 score
192 time
-0.44331626548958036 score
193 time
-0.4786363320973166 score
194 time
-0.4939779291796947 score
195 time
-0.1004798791391641 score
196 time
-0.416961

D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_

207 time
-0.2477342763048238 score
208 time
-0.21799770666831342 score
209 time
-0.2794763347457947 score
210 time
-0.3989890381205383 score
211 time
-0.2632412997401219 score
212 time
-0.24762661119580698 score
213 time
-0.24997291373782304 score
214 time
-0.40065203020009754 score
215 time
-0.40920929870593725 score
216 time
-0.4118624607266408 score
217 time
-0.4044426365643168 score
218 time
-0.4068305301142869 score
219 time
-0.39144433354654445 score
220 time
-0.3821301725292874 score
221 time
-0.3914459179138534 score
222 time
-0.5506205801870847 score
223 time
-0.4488786447541264 score
224 time
-0.546104778642565 score
225 time
-0.5143837253408969 score
226 time
-0.5678221020324639 score
227 time
-0.4113042182044464 score
228 time
-0.4014877359626301 score
229 time
-0.39168561882818875 score
230 time
-0.46509974281932576 score
231 time
-0.4767614678764327 score
232 time
-0.22461293127693446 score
233 time
-0.36679720890426193 score
234 time
-0.36180908449957094 score
235 time
-

D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_


-0.4252161875075066 score
243 time
-0.4294743890463073 score
244 time
-0.4734834341465369 score
245 time
-0.3285948694511429 score
246 time
-0.4379202438239238 score
247 time
-0.38790659537315153 score
248 time
-0.4277684180446051 score
249 time
-0.34126034842582387 score
250 time
-0.29434212836657814 score
251 time
-0.17146143025321314 score
252 time
-0.30312162877191073 score
253 time
-0.3365110908662629 score
254 time
-0.37026092390384857 score
255 time
-0.45795463630181305 score
256 time
-0.3262147430645217 score
257 time
-0.32394839586163915 score
258 time
-0.3049325267315819 score
259 time
-0.26261273215060665 score
260 time
-0.31169388187867164 score
261 time
-0.37477734592742945 score
262 time
-0.4763655002457301 score
263 time
-0.5700259825606344 score
264 time
-0.5731740269145524 score
265 time
-0.5245634277046765 score
266 time
-0.18169912186484938 score
267 time
-0.3916105441977706 score
268 time
-0.4257891215069404 score
269 time
-0.42705367506104985 score
270 time
-0.365

D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_

D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_

313 time
-0.4596949089525699 score
314 time
-0.5330345977026785 score
315 time
-0.5315362164089614 score
316 time
-0.5487946812506876 score
317 time
-0.484255788987751 score
318 time
-0.48410020051687175 score
319 time
-0.45321321062866793 score
320 time
-0.3965840127738718 score
321 time
-0.6435896608055722 score
322 time
-0.6123386779663547 score
323 time
-0.24314688879702806 score
324 time
-0.2829394347783927 score
325 time
-0.26511024464207567 score
326 time
-0.03209700178245134 score
327 time
-0.07029223680748475 score
328 time
-0.17893671550006296 score
329 time
-0.07920096825427048 score
330 time
-0.22428732904678617 score
331 time
-0.18430929020592357 score
332 time
-0.15983805701954523 score
333 time
-0.13677221722060695 score
334 time
-0.1068487017799209 score
335 time
-0.09340341436405564 score
336 time
-0.10049971826305026 score
337 time
-0.18004521001766216 score
338 time
-0.12288692899002517 score
339 time
-0.30302188102064376 score
340 time
-0.21655179714903497 score
341

D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_

352 time
-0.3442534467285332 score
353 time
-0.4697574231611572 score
354 time
-0.46039057869030114 score
355 time
-0.47552889901988293 score
356 time
-0.6413641572531541 score
357 time
-0.6735836530357613 score
358 time
-0.6926581582349539 score
359 time
-0.814485889751929 score
360 time
-0.8217307489951962 score
361 time
-0.8929095833831554 score
362 time
-0.7149855282645866 score
363 time
-0.9751779476192795 score
364 time
-0.8829033316134857 score
365 time
-0.8038504819614645 score
366 time
-0.8373227414008718 score
367 time
-0.7869238976999269 score
368 time
-0.4435697440713888 score
369 time
-0.5817064916390662 score
370 time
-0.26855033020462193 score
371 time
-0.39070340558229133 score
372 time
-0.3697799180987251 score
373 time
-0.45082695053706234 score
374 time
-0.4454293285747972 score
375 time
-0.2089981475552618 score
376 time
-0.224306591279559 score
377 time
-0.17043081200409096 score
378 time
-0.20850870909742716 score
379 time
-0.10930074698743653 score
380 time
-0.14

D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
D:\ANACONDA\envs\tensorflow\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_

-0.18250056542445892 score
385 time
-0.2348743152714443 score
386 time
-0.3113864730531217 score
387 time
-0.29359492780405705 score
388 time
-0.10927218241043124 score
389 time
-0.23086495633932036 score
390 time
-0.22472420604037224 score
391 time
-0.36108175964833045 score
392 time
-0.17030804589778104 score
393 time
-0.3992954039810144 score
394 time
-0.14008252508874475 score
395 time
-0.31063067817359524 score
396 time
-0.0009941125292360908 score
397 time
-0.23661325537863087 score
398 time
-0.266602790473663 score
399 time
-0.010485555741123642 score
400 time
0.03395906043091257 score
401 time
-0.16397455075875156 score
402 time
-0.28624211703668356 score
403 time
-0.1053511056306371 score
404 time
-0.25032581305631063 score
405 time
-0.24107379766834927 score
406 time
-0.256104994897274 score
407 time
0.09877670031423846 score
408 time
-0.056039100910312056 score
409 time
0.048977321493612 score
410 time
-0.13123901234151303 score
411 time
-0.20706313176629654 score
412 time
-

In [454]:
#准备测试数据
df=testlist[0]
test_X=LinearModel_X(df)
test_y=LinearModel_y(df)
test_X=SGD_XScaler.transform(test_X)
test_y=SGD_YScaler.transform(test_y)
pred_y=sgd.predict(test_X)

In [455]:
for i in range(len(pred_y)):
    print(i,test_y[i],pred_y[i])

0 [-1.52133748] -2.0908396333426884
1 [-1.52133748] -2.0908396333426884
2 [-1.52133748] -2.0908396333426884
3 [-1.52133748] -2.133634214169019
4 [-1.51831463] -2.133634214169019
5 [-1.51831463] -2.133634214169019
6 [-1.51831463] -2.133634214169019
7 [-1.51831463] -2.133634214169019
8 [-1.51831463] -2.133634214169019
9 [-1.51831463] -2.133634214169019
10 [-1.51831463] -2.133634214169019
11 [-1.51831463] -2.133634214169019
12 [-1.51831463] -2.133634214169019
13 [-1.51831463] -2.133634214169019
14 [-1.51831463] -2.133634214169019
15 [-1.51831463] -2.133634214169019
16 [-1.51831463] -2.133634214169019
17 [-1.51529177] -2.0834976654014965
18 [-1.51226892] -2.0834976654014965
19 [-1.50924606] -2.0834976654014965
20 [-1.50622321] -2.0834976654014965
21 [-1.50320036] -1.9348222708959886
22 [-1.5001775] -1.9348222708959886
23 [-1.49715465] -1.9348222708959886
24 [-1.49413179] -1.9348222708959886
25 [-1.49413179] -1.9348222708959886
26 [-1.49413179] -1.9348222708959886
27 [-1.49413179] -1.934822

289 [-0.93490381] -1.1226072209248865
290 [-0.93490381] -1.1226072209248865
291 [-0.93490381] -1.1226072209248865
292 [-0.93490381] -1.1226072209248865
293 [-0.93490381] -1.1226072209248865
294 [-0.93490381] -1.1226072209248865
295 [-0.93490381] -1.1226072209248865
296 [-0.93490381] -1.1226072209248865
297 [-0.93490381] -1.1226072209248865
298 [-0.93490381] -1.1226072209248865
299 [-0.93490381] -1.1226072209248865
300 [-0.93490381] -1.1226072209248865
301 [-0.93490381] -1.1226072209248865
302 [-0.93490381] -1.1226072209248865
303 [-0.93490381] -1.1226072209248865
304 [-0.93490381] -1.1226072209248865
305 [-0.93490381] -1.1226072209248865
306 [-0.93490381] -1.1226072209248865
307 [-0.93490381] -1.1226072209248865
308 [-0.93490381] -1.1226072209248865
309 [-0.93490381] -1.1226072209248865
310 [-0.93490381] -1.1226072209248865
311 [-0.93490381] -1.1226072209248865
312 [-0.93490381] -1.1226072209248865
313 [-0.93490381] -1.1226072209248865
314 [-0.93490381] -1.1226072209248865
315 [-0.9349

812 [-0.37869867] -0.5508135324556174
813 [-0.37869867] -0.5508135324556174
814 [-0.37869867] -0.5508135324556174
815 [-0.37869867] -0.5508135324556174
816 [-0.37869867] -0.5508135324556174
817 [-0.37869867] -0.5508135324556174
818 [-0.37869867] -0.5508135324556174
819 [-0.37869867] -0.5508135324556174
820 [-0.37869867] -0.5508135324556174
821 [-0.37869867] -0.5508135324556174
822 [-0.37869867] -0.5508135324556174
823 [-0.37869867] -0.5508135324556174
824 [-0.37869867] -0.5508135324556174
825 [-0.37869867] -0.5508135324556174
826 [-0.37869867] -0.5508135324556174
827 [-0.37869867] -0.5508135324556174
828 [-0.37869867] -0.5508135324556174
829 [-0.37869867] -0.5508135324556174
830 [-0.37869867] -0.5508135324556174
831 [-0.37869867] -0.5508135324556174
832 [-0.37869867] -0.5508135324556174
833 [-0.37869867] -0.5508135324556174
834 [-0.37869867] -0.5508135324556174
835 [-0.37869867] -0.5508135324556174
836 [-0.37869867] -0.5508135324556174
837 [-0.37869867] -0.5508135324556174
838 [-0.3786

1447 [0.65209454] 0.37193599738631344
1448 [0.65209454] 0.37193599738631344
1449 [0.65209454] 0.37193599738631344
1450 [0.65209454] 0.37193599738631344
1451 [0.65209454] 0.37193599738631344
1452 [0.65209454] 0.37193599738631344
1453 [0.65209454] 0.37193599738631344
1454 [0.65209454] 0.37193599738631344
1455 [0.65209454] 0.37193599738631344
1456 [0.65209454] 0.37193599738631344
1457 [0.65209454] 0.37193599738631344
1458 [0.65209454] 0.37193599738631344
1459 [0.65209454] 0.37193599738631344
1460 [0.65209454] 0.37193599738631344
1461 [0.65209454] 0.37193599738631344
1462 [0.65209454] 0.37193599738631344
1463 [0.65209454] 0.37193599738631344
1464 [0.65209454] 0.37193599738631344
1465 [0.65209454] 0.37193599738631344
1466 [0.65209454] 0.37193599738631344
1467 [0.65209454] 0.37193599738631344
1468 [0.65209454] 0.37193599738631344
1469 [0.65209454] 0.37193599738631344
1470 [0.65209454] 0.37193599738631344
1471 [0.65209454] 0.37193599738631344
1472 [0.65209454] 0.37193599738631344
1473 [0.6520

1981 [1.52569934] 0.9494094662230059
1982 [1.5287222] 0.9494094662230059
1983 [1.53174505] 0.9494094662230059
1984 [1.5347679] 0.9494094662230059
1985 [1.53779076] 0.9494094662230059
1986 [1.54081361] 0.9494094662230059
1987 [1.54383647] 0.9494094662230059
1988 [1.54685932] 0.9494094662230059
1989 [1.54988217] 0.9494094662230059
1990 [1.55290503] 0.9494094662230059
1991 [1.55592788] 0.9494094662230059
1992 [1.56197359] 0.9494094662230059
1993 [1.56499644] 0.9494094662230059
1994 [1.5680193] 0.9494094662230059
1995 [1.57104215] 0.9995460149905291
1996 [1.57406501] 0.9995460149905291
1997 [1.58011071] 0.9995460149905291
1998 [1.58011071] 0.9995460149905291
1999 [1.58313357] 0.9995460149905291
2000 [1.58313357] 0.9995460149905291
2001 [1.58615642] 0.9995460149905291
2002 [1.58917928] 0.9995460149905291
2003 [1.59220213] 0.9995460149905291
2004 [1.59522498] 0.9995460149905291
2005 [1.59824784] 0.9995460149905291
2006 [1.60127069] 0.9995460149905291
2007 [1.60429355] 0.9995460149905291
2008